src/script_lua.c

```diff
@@ -1709,3 +1692,18 @@ void luaCallFunction(scriptRunCtx* run_ctx, lua_State *lua, robj** keys, size_t
 unsigned long luaMemory(lua_State *lua) {
     return lua_gc(lua, LUA_GCCOUNT, 0) * 1024LL;
 }
+
+/* Call the Lua garbage collector from time to time to avoid a
+ * full cycle performed by Lua, which adds too latency.
+ *
+ * The call is performed every LUA_GC_CYCLE_PERIOD executed commands
+ * (and for LUA_GC_CYCLE_PERIOD collection steps) because calling it
+ * for every command uses too much CPU. */
+#define LUA_GC_CYCLE_PERIOD 50
+void luaGC(lua_State *lua, int *gc_count) {
+    (*gc_count)++;
```

Comment:
did we mean that gc_count is unique per lua_State? (i.e. each state has one counter).
if one state has two counters, each of them acts as a separate trigger, but the memory that's collected an the triggers for garbage accumulation is shared.

maybe each sub-system (functions / eval) should have it's own `luaGC` function with a static variable as counter inside?

Reply:
> So the general idea is that the same trigger we had per script invocation, is now added when a script is loaded, and to do that some small refactoring was needed. is that right?

yes.

> did we mean that gc_count is unique per lua_State? (i.e. each state has one counter).

It should be that both lua script and lua function have their own unique counter.

> maybe each sub-system (functions / eval) should have it's own luaGC function with a static variable as counter inside?

i think the commit before [`d348d8c` (#13407)](https://github.com/redis/redis/pull/13407/commits/d348d8c5846e4abff437e7bbc5d376b3b08bafbc) is as you say, but i'll be a duplicate of luaGC code, so i just created separate static `gc_count` for both of them.

Reply:
ohh, for some reason i thought you now have 3 counters. i.e. that one state has two separate counters.
well, both approaches are ok.
please try to improve the inline comments and the PR description to make sure the points in this discussion thread are clear.